aioxmpp/avatar/service.py

```diff
@@ -391,17 +440,146 @@ class AvatarService(service.Service):
 
     def __init__(self, client, **kwargs):
         super().__init__(client, **kwargs)
-        self._has_pep = None
-        self._metadata_cache = {}
+        self._has_pep_avatar = set()
+        self._metadata_cache = LRUDict()
+        self._metadata_cache.maxsize = 200
         self._pubsub = self.dependencies[pubsub.PubSubClient]
-        self._notify_lock = asyncio.Lock()
+        self._pep = self.dependencies[pep.PEPClient]
+        self._presence_server = self.dependencies[presence.PresenceServer]
         self._disco = self.dependencies[disco.DiscoClient]
+        self._vcard = self.dependencies[vcard.VCardService]
         # we use this lock to prevent race conditions between different
         # calls of the methods by one client.
         # XXX: Other, independent clients may still cause inconsistent
         # data by race conditions, this should be fixed by at least
         # checking for consistent data after an update.
         self._publish_lock = asyncio.Lock()
+        self._synchronize_vcard = False
+        self._vcard_ressource_interference = set()
+        self._vcard_id = None
+        self._vcard_rehash_task = None
+
+    @property
+    def metadata_cache_size(self):
+        """
+        Maximum number of cache entries in the avatar metadata cache.
+
+        This is mostly a measure to prevent malicious peers from
+        exhausting memory by spamming vCard based avatar metadata for
+        different ressources.
+
+        .. versionadded:: 0.10
+
+        """
+        return self._metadata_cache.maxsize
+
+    @metadata_cache_size.setter
+    def metadata_cache_size(self, value):
+        self._metadata_cache.maxsize = value
+
+    @property
+    def synchronize_vcard(self):
+        """
+        Set this property to true to enable publishing the a vCard avatar.
+
+        This property defaults to false. For the setting true to have
```

Comment:
Any technical reason to default to false?

Reply:
Primarily aesthetic reasons (don't actively use the legacy protocol by default)

Reply:
Ejabberd would have to be able to work with that, it’s not the clients fault in that case (if they need special casing, they should negotiate that feature).

I don’t have strong opinions either way, in any case we’ll just enable it in JabberCat.